### PR TITLE
[RFC] ASoC: SOF: set link dma channel in hda config

### PR DESCRIPTION
**Pull request by RanderWang**
Now FW set wrong link dma channel which make capture failed.
My algorithm is: allocate link dma channel in host and
send hda config to FW to config link dma.

This is a prototype, please ignore the format. Please give me some advices, thanks!

Signed-off-by: Rander Wang <rander.wang@linux.intel.com>

### CODE REVIEW

**Comment by RanderWang** [comment]
In FW, there are HDA playback dai and capture dai, because they use different DMA engine.
So, I need to config dai according to capability

File: sound/soc/sof/intel/cnl.c [hunk 1]
```diff
@@ -230,6 +230,8 @@ struct snd_sof_dsp_ops sof_cnl_ops = {
 	.pcm_hw_params	= hda_dsp_pcm_hw_params,
 	.pcm_trigger	= hda_dsp_pcm_trigger,
 
+	.dai_config	= hda_dai_config,
```
Comment (plbossart) [inline sound/soc/sof/intel/cnl.c]: is this really specific to cnl?
How would you handle the case of a platform which can work in either I2S or HDaudio modes, with a run-time selection done?
and does this really belong in the ops or should it be a hook private to HDaudio cases?

File: include/sound/sof/dai-intel.h [hunk 2]
```diff
@@ -83,7 +83,7 @@ struct sof_ipc_dai_ssp_params {
 /* HDA Configuration Request - SOF_IPC_DAI_HDA_CONFIG */
 struct sof_ipc_dai_hda_params {
 	struct sof_ipc_hdr hdr;
-	/* TODO */
```
Comment (plbossart) [inline include/sound/sof/dai-intel.h:86]: we need to remove the hdr, it's not needed. Maybe do this in a separate patch for API change on firmware and driver at the same time.

File: sound/soc/sof/intel/hda-stream.c [hunk 3]
```diff
@@ -497,6 +497,9 @@ int hda_dsp_stream_init(struct snd_sof_dev *sdev)
 	num_playback = (gcap >> 12) & 0x0f;
 	num_total = num_playback + num_capture;
 
+	sdev->num_playback = num_playback;
+	sdev->num_capture = num_capture;
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-stream.c]: what happens if we have a platform with HDaudio and DMICs? They are already on the market for WHL windows platforms, need to think of this case upfront.

File: sound/soc/sof/intel/hda-dai.c [hunk 4]
```diff
@@ -33,6 +33,43 @@ struct hda_pipe_params {
 	unsigned int link_bps;
 };
 
+int hda_dai_config(struct snd_sof_dev *sdev,
+					struct snd_soc_dai_link *link,
+					struct sof_ipc_dai_config *config,
+					int direction)
+{
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dai.c:47]: alignment is out of whack?

File: sound/soc/sof/intel/hda-dai.c [hunk 5]
```diff
@@ -33,6 +33,43 @@ struct hda_pipe_params {
...
+	struct snd_pcm_substream substream;
+	struct snd_soc_dai *dai;
+	struct hdac_ext_stream *stream;
+	struct snd_soc_dai_link_component dai_component = {0};
+	struct hdac_bus *bus = sof_to_bus(sdev);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dai.c]: init needed?
Also reverse xmas tree order.

File: sound/soc/sof/ops.h [hunk 6]
```diff
@@ -312,6 +312,18 @@ snd_sof_pcm_platform_pointer(struct snd_sof_dev *sdev,
...
 }
 
+static inline int
+snd_sof_dai_config(struct snd_sof_dev *sdev,
+			struct snd_soc_dai_link *link,
+			struct sof_ipc_dai_config *config,
+			int direction)
+{
+	 if (sdev->ops && sdev->ops->dai_config)
+		 return sdev->ops->dai_config(sdev, link, config, direction);
```
Comment (plbossart) [inline sound/soc/sof/ops.h]: I am not sure this is right? We will have multiple dais, some of which will need a config and others will not. The level of indirection should be at the dai level and probably part of the hw_params?

File: sound/soc/sof/topology.c [hunk 7]
```diff
@@ -2057,14 +2057,31 @@ static int sof_link_hda_load(struct snd_soc_component *scomp, int index,
 	dev_dbg(sdev->dev, "tplg: config HDA%d fmt 0x%x\n",
 		config->dai_index, config->format);
 
-	/* send message to DSP */
-	ret = sof_ipc_tx_message(sdev->ipc,
-				 config->hdr.cmd, config, size, &reply,
-				 sizeof(reply));
+	if (link->dpcm_playback) {
+		snd_sof_dai_config(sdev, link, config, SNDRV_PCM_STREAM_PLAYBACK);
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: comment related to the others, is it needed to have an sdev-level indirection when this is only called in the case of HDaudio and on a per-DAI basis? It feels like a private hda-specific case, not a global callback.

File: sound/soc/sof/intel/hda-dai.c (hunk 4 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c:47]: yes, I will refine it later. I just finish this RFC before leaving office. Sorry, it is not perfect.

File: sound/soc/sof/intel/hda-dai.c (hunk 5 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c]: yes, I will refine it.

File: sound/soc/sof/topology.c (hunk 7 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c]: yes, It is ugly for a issue in FW, but later I find another way to do it. I will refine it

File: sound/soc/sof/intel/cnl.c (hunk 1 above)
Comment (RanderWang) [reply sound/soc/sof/intel/cnl.c]: yes, APL and ICL should be included.

I will add a macro ifdef _HDA_ to control it

**Comment by RanderWang** [comment]
Update my PR. This patch is not only for capture but also playback. Now it is lucky to get playback work.
Capture is failed by DMA trace which uses dma channel zero just like a capture stream. so this make a result that host sets dma channel 0 but FW has to use dma channel 1.

File: sound/soc/sof/ops.h (hunk 6 above)
Comment (RanderWang) [reply sound/soc/sof/ops.h]: We made a agreement that it should be done as SSP at config time.
My last PR is done at hw_params. But it doesn't work for tone because there maybe
no hw_params for tone.

File: sound/soc/sof/intel/cnl.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/intel/cnl.c]: Humm, not sure. Even if you have HDA configured, you may have DMICs. So not all DAIs will support this callback. and since you invoke the callback from HDaudio code only, this might be better handled from a HDaudio-specific ops or helper.

File: sound/soc/sof/ops.h (hunk 6 above)
Comment (plbossart) [reply sound/soc/sof/ops.h]: @ranj063 can you comment on the tone? I am pretty sure you did a modification to pass the hw_params.

**Comment by plbossart** [comment]
> Update my PR. This patch is not only for capture but also playback. Now it is lucky to get playback work.
> Capture is failed by DMA trace which uses dma channel zero just like a capture stream. so this make a result that host sets dma channel 0 but FW has to use dma channel 1.

Do you mean to say playback works because the link ID is set to zero which happens to be the one used by firmware?

**Comment by RanderWang** [comment]
> > Update my PR. This patch is not only for capture but also playback. Now it is lucky to get playback work.
> > Capture is failed by DMA trace which uses dma channel zero just like a capture stream. so this make a result that host sets dma channel 0 but FW has to use dma channel 1.
> 
> Do you mean to say playback works because the link ID is set to zero which happens to be the one used by firmware?

Your description is for capture because of DMA trace.  For playback, host setting matches FW setting.

File: sound/soc/sof/ops.h (hunk 6 above)
Comment (RanderWang) [reply sound/soc/sof/ops.h]: @ranj063  yes, I am also very curious about this. I got some info from keyon only

File: sound/soc/sof/intel/hda-stream.c (hunk 3 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-stream.c]: It is removed

File: sound/soc/sof/intel/cnl.c (hunk 1 above)
Comment (RanderWang) [reply sound/soc/sof/intel/cnl.c]: yes, I made  a change

File: sound/soc/sof/intel/cnl.c [hunk 8]
```diff
@@ -188,6 +188,29 @@ static int cnl_ipc_send_msg(struct snd_sof_dev *sdev,
...
+		break;
+	default:
+		dev_err(sdev->dev, "error: invalid DAI type %d\n", config->type);
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
```
Comment (plbossart) [inline sound/soc/sof/intel/cnl.c]: Meh. copy/paste code isn't really accepted in Linux circles.

File: sound/soc/sof/intel/cnl.c [hunk 9]
```diff
@@ -230,6 +253,8 @@ struct snd_sof_dsp_ops sof_cnl_ops = {
 	.pcm_hw_params	= hda_dsp_pcm_hw_params,
 	.pcm_trigger	= hda_dsp_pcm_trigger,
 
+	.dai_fixup	= cnl_dai_fixup,
+
```
Comment (plbossart) [inline sound/soc/sof/intel/cnl.c]: does this belong in the SOF ops, or could you use the existing ASoC .fixup?

File: sound/soc/sof/ops.h (hunk 6 above)
Comment (ranj063) [reply sound/soc/sof/ops.h]: @RanderWang @plbossart sorry I havent had the time to review this PR yet. But seems like the code is outdated now and I'm lost on what the original question was about.

Comment (plbossart) [reply sound/soc/sof/ops.h]: the question was about tone support requiring hw_params to be set even in the absence of a PCM FE.

Comment (ranj063) [reply sound/soc/sof/ops.h]: @plbossart @RanderWang yes, the tone component didnt work for the longest time because I wasnt setting hw_params. So in order to set the hw_params, we introduced the concept of a virtual PCM FE (a pcm that is not registered with ALSA) which would be used to set the hw_params.

Comment (keyonjie) [reply sound/soc/sof/ops.h]: @ranj063 do you have chance to talk about this(virtual PCM) in alsa-devel mailing list? That's a good idea to handle kinds of problems like this tone one, it would be good if we can align this solution from Takashi and Mark before implement SOF specific coding.

File: sound/soc/sof/intel/cnl.c (hunk 8 above)
Comment (RanderWang) [reply sound/soc/sof/intel/cnl.c]: yes, I will refine it later

File: sound/soc/sof/intel/cnl.c (hunk 9 above)
Comment (RanderWang) [reply sound/soc/sof/intel/cnl.c]: I check the fixup in ASoc. It is not fit us. for three reasons
(1) the input params are not fit for link_hda_load
(2) the link dma channel is not returned by these functions
(3) these functions are called at the time of hw_param

As you know, hda config is set at core layer and hda dma channel info could be only gotten at intel layer. It is a trouble I get, so I add a ops to get hda info from Intel layer. And I think other vendors may also want a ops to get platform info if they use SOF.

If it is not good, how about my last RFC about this, It is done in hw_params. If it is a candidate, I will refine it later.

File: sound/soc/sof/ops.h (hunk 6 above)
Comment (ranj063) [reply sound/soc/sof/ops.h]: @keyonjie not yet. I'll have to re-visit the tone patches soon.

**Comment by ranj063** [comment]
> > Update my PR. This patch is not only for capture but also playback. Now it is lucky to get playback work.
> > Capture is failed by DMA trace which uses dma channel zero just like a capture stream. so this make a result that host sets dma channel 0 but FW has to use dma channel 1.
> 
> Do you mean to say playback works because the link ID is set to zero which happens to be the one used by firmware?

@plbossart after my chat with Rander I understood the problem as follows:
Capture is because trace dma acquires channel 0 when it is initialized.
And then when link dma ch is assigned for capture, it happens to assign channel 0 too. And we assume that we should use the same channel for link dma and host dma thereby clashing with trace dma.

@RanderWang please correct me if I am wrong

@ranj063 
[Rander] The sorry is a little complex. Because trace dma use channel 0, host dma has to use channel 1 and link dma uses the setting of host dma. But host driver get stream 0 , set stream 0 and expect FW
to use link dma channel 0. So it is a mismatch between host and FW.

**Comment by keyonjie** [comment]
Sorry to chime in late, after discussion with Rander about the implementation difficulties with assigning Link DMA channels at topology parsing stage, I suggest we switch back to previous solution, that is, sending link_dma_id in hw_params(), and Ranjani's explanation about virtual PCM and hw_params() usage in tone case give me more confidence about doing it in that way:

in driver side,
we will assign link_dma_ch in BE dai(hda_dai) hw_params(), and then pass this link_dma_ch to FW in FE PCM hw_params().

in FW side,
today, we actually get link_dma channel at hw_params() stage, so we only need a minor change where we get link_dma_chan from, no process changes needed.

they are both minor changes with low risk to me, I vote for it.

**Comment by ranj063** [comment]
@keyonjie for my benefit :
When does hw_params for BE dai get called?

Is that something that happens only once?

**Comment by RanderWang** [comment]
@ranj063   please refer to
dpcm_fe_dai_hw_params() {
     dpcm_be_dai_hw_params();
     soc_pcm_hw_params();
}

According to ASoC, it is called by fe_dai

**Comment by RanderWang** [comment]
@ranj063  if one BE is shared by multiple FEs, BE hw_params would be called multiple times. Each FE would call BE hw_params. Just think that each FE works randomly not at the same time regularly, so each FE need to set BE.

**Comment by plbossart** [comment]
> in driver side,
> we will assign link_dma_ch in BE dai(hda_dai) hw_params(), and then pass this link_dma_ch to FW in FE PCM hw_params().

This is **NOT** a good design. You can have multiple FEs connected to a BE, and conversely we could send the same data on multiple BEs (think playing an alarm on both speakers and headset). We've circled on this before, why are we going back on this? This is a DAI property, not a FE property.

File: sound/soc/sof/intel/cnl.c (hunk 9 above)
Comment (plbossart) [reply sound/soc/sof/intel/cnl.c]: Sorry @RanderWang , I don't get your answers.
1) If you look at the SST and machine drivers, they've used the tx- and rx-masks for stream tags, so why are the parameters for fixup not right? It may not be elegant to abuse a definition but it's been done before.
2) why would the DMA channel be returned by the fixup when you want to pass it to firmware using the fixup?
3) to paraphrase your answer, the functions are called during hw_params, and your last counter proposal is to say "how about my last RFC about this, It is done in hw_params". there is a serious logic issue here that I can't sort out without clarifications.

**Comment by ranj063** [comment]
@plbossart a few more questions for my understanding: 
1. Is link dma used in the case of HDMI playback?
2. If so, we do we use the decoupled mode in this case? 
Thanks!

**Comment by plbossart** [comment]
>  1. Is link dma used in the case of HDMI playback?
>
yes, same as for HDaudio
>
>  1. If so, we do we use the decoupled mode in this case?
>
use, there is no other choice. The only things that work are coupled 
(for legacy w/o DSP) and decoupled (with DSP managing the link DMA).
>
> 1.
>

**Comment by ranj063** [comment]
Thanks @plbossart. Then does it make sense to introduce the notion of coupled/decouple in the link dma config? Then we could choose to use the stream_tag or the link_dma_ch during dai_params depending on the mode.

**Comment by keyonjie** [comment]
> > in driver side,
> > we will assign link_dma_ch in BE dai(hda_dai) hw_params(), and then pass this link_dma_ch to FW in FE PCM hw_params().
> 
> This is **NOT** a good design. You can have multiple FEs connected to a BE, and conversely we could send the same data on multiple BEs (think playing an alarm on both speakers and headset). We've circled on this before, why are we going back on this? This is a DAI property, not a FE property.

We can add an IPC type for this 'DAI params', e.g. SOF_IPC_DAI_PARAMS, we will assign link_dma_ch in BE dai(hda_dai) hw_params(), and send SOF_IPC_DAI_PARAMS to FW, the bonus of adding this is that it's possible to send dynamic hw_params for BE DAIs(including SSP DAI), comparing to that hard-coded in topology previously.

File: sound/soc/sof/intel/cnl.c (hunk 9 above)
Comment (RanderWang) [reply sound/soc/sof/intel/cnl.c]: @plbossart  
(1) ok, I get your idea.  
(2) The function is called dai_config, and was changed to fixup.
(3) please ignore my question

**Comment by RanderWang** [comment]
test @RanderWang

**Comment by RanderWang** [comment]
Update my RFC.
Don't define more ops to get hda info, reuse the interface of ASoC

File: sound/soc/sof/intel/hda-dai.c (hunk 4 above)
Comment (lgirdwood) [reply sound/soc/sof/intel/hda-dai.c:47]: Also better if the masks and slots can be embedded in a new structure, that way you are only passing a single ptr to a structure rather than 4 ptrs to ints.

File: sound/soc/sof/topology.c [hunk 10]
```diff
@@ -2028,16 +2028,56 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
 	return ret;
 }
 
+static inline int sof_link_hda_process(struct snd_sof_dev *sdev,
+				       struct snd_soc_dai_link *link,
+				       struct sof_ipc_dai_config *config,
+					      int slot,
+				       int direction)
```
Comment (lgirdwood) [inline sound/soc/sof/topology.c]: alignment.

File: sound/soc/sof/topology.c [hunk 11]
```diff
@@ -2054,18 +2094,47 @@ static int sof_link_hda_load(struct snd_soc_component *scomp, int index,
...
+		dev_err(sdev->dev, "failed to find dai %s", dai->name);
+		return -EINVAL;
+	}
 
-	/* send message to DSP */
-	ret = sof_ipc_tx_message(sdev->ipc,
-				 config->hdr.cmd, config, size, &reply,
-				 sizeof(reply));
+	tx_num = link->dpcm_playback;
+	rx_num = link->dpcm_capture;
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: if I am not mistaken, dpcm_playback and dpcm_capture are just booleans. using tx/rx_num hints at multiple streams so it's confusing.

File: sound/soc/sof/topology.c [hunk 12]
```diff
@@ -2054,18 +2094,47 @@ static int sof_link_hda_load(struct snd_soc_component *scomp, int index,
...
 
-	if (ret < 0)
-		dev_err(sdev->dev, "error: failed to set DAI config for HDA%d\n",
+	ret = snd_soc_dai_get_channel_map(dai, &tx_num, &tx_slot,
+					  &rx_num, &rx_slot);
+	if (ret < 0) {
```
Comment (plbossart) [inline sound/soc/sof/topology.c:2121]: beyond the naming issue, tx_num and rx_num are not changed so no need to pass them as pointers, or you are not using the API in the expected way?

File: sound/soc/sof/topology.c (hunk 10 above)
Comment (plbossart) [reply sound/soc/sof/topology.c]: it's way too big for an inline function?

File: sound/soc/sof/topology.c [hunk 13]
```diff
@@ -2054,18 +2094,47 @@ static int sof_link_hda_load(struct snd_soc_component *scomp, int index,
...
+			return ret;
+		}
+	}
+
+	if (link->dpcm_capture) {
+		ret = sof_link_hda_process(sdev, link, config, rx_slot,
+					   SNDRV_PCM_STREAM_CAPTURE);
+		if (ret < 0)
+			dev_err(sdev->dev, "error: failed to set DAI config for HDA%d\n",
+				config->dai_index);
```
Comment (plbossart) [inline sound/soc/sof/topology.c:2147]: use different error message for capture and playback.

File: sound/soc/sof/topology.c [hunk 14]
```diff
@@ -2028,16 +2028,56 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
...
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_soc_dai_link_component dai_component = {0};
 	struct snd_soc_tplg_private *private = &cfg->priv;
-	struct sof_ipc_reply reply;
+	struct snd_soc_dai *dai;
+
 	u32 size = sizeof(*config);
+	u32 tx_num = 0, tx_slot;
+	u32 rx_num = 0, rx_slot;
 	int ret;
```
Comment (plbossart) [inline sound/soc/sof/topology.c:2090]: i like one variable per line, otherwise inits are hidden in plain sight...

File: sound/soc/sof/topology.c [hunk 15]
```diff
@@ -2028,16 +2028,56 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
...
+		if (!dai->name)
+			continue;
+
+		if (strcmp(link->name, dai->name) == 0 &&
+		    dai->comp_dai.direction == direction) {
+			config->dai_index = dai->comp_dai.dai_index;
+			break;
+		}
+	}
+
```
Comment (plbossart) [inline sound/soc/sof/topology.c:2058]: check that there was a match in the list.

File: sound/soc/sof/intel/hda-dai.c [hunk 16]
```diff
@@ -33,7 +33,57 @@ struct hda_pipe_params {
...
 
-/* TODO: add hda dai params in tplg, and configure this in topology parsing */
+/* Unlike GP dma, there is a set of stream registers in hda controller to
+ * control the link dma channels. Each register controls one link dma
+ * channel and the relation is fixed. To make sure FW uses the correct
+ * link dma channel, host allocates stream register and sends the
+ * corresponding link dma channel to FW to allocate link dma channel
+ */
+static int hda_link_get_channel_map(struct snd_soc_dai *dai,
+				    unsigned int *tx_num, unsigned int *tx_slot,
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dai.c]: careful. channel_map is typically used for PCM data. You want to use something like hda_get_link_dma_channel_map
Also wondering if you need to deallocate the stream to release all resources?

File: sound/soc/sof/topology.c (hunk 11 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c]: yes, I made a tick. I will do it if(link->dpcm_playback) tx_num = 1.

File: sound/soc/sof/topology.c (hunk 13 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c:2147]: yes, I will

File: sound/soc/sof/topology.c (hunk 14 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c:2090]: yes

File: sound/soc/sof/topology.c (hunk 15 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c:2058]: yes, I will

File: sound/soc/sof/intel/hda-dai.c (hunk 4 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c:47]: this is a ASoC interface and I just reuse it for hda.

File: sound/soc/sof/topology.c (hunk 10 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c]: yes, I will refine it

File: sound/soc/sof/topology.c (hunk 15 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c]: yes, I will

File: sound/soc/sof/topology.c (hunk 12 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c:2121]: yes, I use it special for hda link. machine driver may create a dai_link only for playback or capture, so I use tx_num to check whether playback is supported, rx_num for capture

File: sound/soc/sof/intel/hda-dai.c (hunk 16 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c]: yes, good point. These resource are deallocated at hw_free. But I find that hw_free is called after playback or capture stop. This behavior is not we want. So I will refine hw_free and find another way to free it.

**Comment by ranj063** [comment]
@plbossart @RanderWang I am skeptical about a few things with the use of get_channel_map API: 

1. The definition of the API talks about setting up the DAI for TDM operation. Is this relevant for HDA codecs? 
2. If it is indeed OK, why use get_channel_map() when you are infact doing a set_channel_map()?

Dont mean to be hindering progress but this will help me educate myself too.

**Comment by RanderWang** [comment]
> @plbossart @RanderWang I am skeptical about a few things with the use of get_channel_map API:
> 
>     1. The definition of the API talks about setting up the DAI for TDM operation. Is this relevant for HDA codecs?
> 
>     2. If it is indeed OK, why use get_channel_map() when you are infact doing a set_channel_map()?
> 
> 
> Dont mean to be hindering progress but this will help me educate myself too.

I search get_channel_map in Linux kernel, and get no reference of it. So I reuse it . I need a channel allocated by hda, not set a channel to hda, so I use get_channel_map. And I add a set_channel_map to release resource.

File: sound/soc/sof/topology.c [hunk 17]
```diff
@@ -2152,10 +2233,75 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
...
+	if (link->dpcm_playback)
+		tx_num = 1;
+
+	if (link->dpcm_capture)
+		rx_num = 1;
+
+	ret = snd_soc_dai_set_channel_map(dai, tx_num, NULL,
+					  rx_num, NULL);
```
Comment (lyakh) [inline sound/soc/sof/topology.c]: You could just do
ret = snd_soc_dai_set_channel_map(dai, link->dpcm_playback, NULL,
					  link->dpcm_capture, NULL);
and eliminate rx_num and tx_num completely in this function

Comment (RanderWang) [reply sound/soc/sof/topology.c]: link->dpcm_playback is defined as bool variable, so I don't do like this.

**Comment by plbossart** [comment]
> > @plbossart @RanderWang I am skeptical about a few things with the use of get_channel_map API:
> > ```
> > 1. The definition of the API talks about setting up the DAI for TDM operation. Is this relevant for HDA codecs?
> > 
> > 2. If it is indeed OK, why use get_channel_map() when you are infact doing a set_channel_map()?
> > ```
> > Dont mean to be hindering progress but this will help me educate myself too.
> 
> I search get_channel_map in Linux kernel, and get no reference of it. So I reuse it . I need a channel allocated by hda, not set a channel to hda, so I use get_channel_map. And I add a set_channel_map to release resource.

Ranjani is correct though that you are not really using get_channel_map in the intended way, since you ignore tx_num and rx_num. And along the same lines, you assign a number of resources in the _get but the _set is mainly about double-checking for errors. I am not sure this is consistent, you may want to allocated a stream_tag in the get if there is none, and do all the actions with side-effects in the _set.

It may also be that you are doing an analogy between get/put (as done for runtime_pm) and get/set (as done for general controls and usual resources access), but they have vastly different semantics. A set is not supposed to release resources in general.

**Comment by RanderWang** [comment]
I refine my patch, Now release hda stream by hw_free interface in dai driver.

**Comment by ranj063** [comment]
@RanderWang looks good but for the fact that I think we still need to use set_channel_map instead of get_channel_map. 

BTW isthere a FW patch that goes with this PR? I'd like to give it a try on my board.

**Comment by RanderWang** [comment]
update patch in FW

File: sound/soc/sof/topology.c [hunk 18]
```diff
@@ -2152,10 +2233,67 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
...
+	dai = snd_soc_find_dai(&dai_component);
+	if (!dai) {
+		dev_err(sdev->dev, "failed to find dai %s", dai->name);
+		return -EINVAL;
+	}
+
+	if (dai->driver->ops->hw_free)
+		ret = dai->driver->ops->hw_free(NULL, dai);
+	if (ret < 0)
+		dev_err(sdev->dev, "error: failed to set dma channel for %s\n",
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: the error message seems wrong?

File: sound/soc/sof/topology.c [hunk 19]
```diff
@@ -2152,10 +2233,67 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
...
+	}
+
+	if (!sof_dai) {
+		dev_err(sdev->dev, "failed to find dai %s", link->name);
+		return -EINVAL;
+	}
+
+	switch (sof_dai->dai_config.type) {
+	case SOF_DAI_INTEL_SSP:
+	case SOF_DAI_INTEL_DMIC:
```
Comment (plbossart) [inline sound/soc/sof/topology.c:2284]: ```` /* fall through */```` comment needed

File: sound/soc/sof/intel/hda-dai.c [hunk 20]
```diff
@@ -147,22 +194,49 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
...
+		name = rtd->codec_dai->component->name;
+		link = snd_hdac_ext_bus_get_link(bus, name);
+		if (!link)
+			return -EINVAL;
+
+		stream_tag = hdac_stream(link_dev)->stream_tag;
+		snd_hdac_ext_link_clear_stream_id(link, stream_tag);
+
+		link_dev->link_prepared = 0;
+	} else {
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-dai.c:222]: in what cases is the 'else' selected? I really don't understand this code, sorry.

Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c:222]: else case: pcm is NULL, this the case for link dma resource free.

File: sound/soc/sof/topology.c (hunk 19 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c:2284]: yes, I will add it

File: sound/soc/sof/topology.c (hunk 18 above)
Comment (RanderWang) [reply sound/soc/sof/topology.c]: yes, I will refine it

File: sound/soc/sof/intel/hda-dai.c (hunk 20 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-dai.c:222]: let me re-ask the question: is hda_link_hw_free() called at different times with substream==NULL and substream != NULL conditions? if yes, can you explain the two cases?

**Comment by mengdonglin** [comment]
@plbossart @RanderWang Lack of setting link DMA channel in HD-A config in kernel side caused regression https://github.com/thesofproject/linux/issues/322

Since the firmware has merged thesofproject/sof/pull/634 which requires this feature.

**Comment by plbossart** [comment]
> @plbossart @RanderWang Lack of setting link DMA channel in HD-A config in kernel side caused regression #322
> 
> Since the firmware has merged [thesofproject/sof/pull/634](https://github.com/thesofproject/sof/pull/634) which requires this feature.

What can I say? I've mentioned multiple times that we should not merge IPC changes on one side only. I am not going to merge code that is still being reviewed, sorry.

**Comment by RanderWang** [comment]
> > @plbossart @RanderWang Lack of setting link DMA channel in HD-A config in kernel side caused regression #322
> > Since the firmware has merged [thesofproject/sof/pull/634](https://github.com/thesofproject/sof/pull/634) which requires this feature.
> 
> What can I say? I've mentioned multiple times that we should not merge IPC changes on one side only. I am not going to merge code that is still being reviewed, sorry.

@plbossart   It is not caused by IPC change, as you know this IPC change doesn't make any non-hda case failed because it is not change the size of IPC package (union) and is not used now. Please refer to our CI test, all cases pass!

Just HDA playback is failed.

File: sound/soc/sof/intel/hda-dai.c (hunk 20 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c:222]: yes, two cases:
(1) it is called when audio stream is stopped, at this time you could get substream to do some work
(2) at dai_link_unload. I reuse this function with null substream.
It is very tricky to do like this. I also don't like it now. Now I want to do it in hda_dsp_free, but it is impossible do at there because there is no dai.

I am very struggled to find a interface to free logical stream.  Actually, it is no need to free it and it would be free when sof is removed.

**Comment by mengdonglin** [comment]
@plbossart I'm not pushing you to merge the PR since it seems we need more time to refine the solution kernel side.

@lgirdwood It seems we need to revert the thesofproject/sof/pull/634  from FW master branch, to solve the toplogy loading failure https://github.com/thesofproject/linux/issues/322.

**Comment by wenqingfu** [comment]
> I will add follow-up comments myself stating that this will be reworked.

Can you please add an enhancement ticket to track it? @plbossart

**Comment by RanderWang** [comment]
@plbossart  thanks for your guide, it is a very good lesson for me. I will go on moving forward.
